apiserver/facades/agent/uniter/uniter.go

```diff
@@ -406,15 +386,32 @@ func (u *UniterAPI) AvailabilityZone(ctx context.Context, args params.Entities)
 			results.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
 			continue
 		}
-		err = apiservererrors.ErrPerm
-		if canAccess(tag) {
-			var zone string
-			zone, err = getZone(ctx, u.st, u.machineService, tag)
-			if err == nil {
-				results.Results[i].Result = zone
-			}
+
+		if !canAccess(tag) {
+			results.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
 		}
-		results.Results[i].Error = apiservererrors.ServerError(err)
+
+		unitName, err := coreunit.NewName(tag.Id())
+		if err != nil {
+			results.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		machineUUID, err := u.applicationService.GetUnitMachineUUID(ctx, unitName)
+		if err != nil {
+			results.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		az, err := u.machineService.AvailabilityZone(ctx, machineUUID.String())
```

Comment:
Whilst you're here can you make machine service take a `machine.UUID` and not a string.

Reply:
I'll do this in a separate PR.